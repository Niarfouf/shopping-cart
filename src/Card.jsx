export default function Card({ title, price, description, imageUrl }) {
    return (
        <>
            <div className="card">
                <img className="item-image" alt={title} src={imageUrl}></img>
                <h3 className="title">{title}</h3>
                <p>{description}</p>
                <input type="number"></input>
                <button>Add</button>
                <p>{price}</p>
            </div>
        </>
    )
}
