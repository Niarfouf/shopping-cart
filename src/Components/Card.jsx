import { useEffect, useState } from 'react'
export default function Card({ item, addItemInList, quantity = 0, action }) {
    const [input, setInput] = useState(0)
    function left() {
        setInput((input) => (input > 0 ? input - 1 : input))
    }
    function right() {
        setInput((input) => input + 1)
    }
    useEffect(() => {
        setInput(quantity)
    }, [quantity])
    return (
        <>
            <div className="card">
                <img
                    className="item-image"
                    alt={item.title}
                    src={item.image}
                ></img>
                <h3 className="title">{item.title}</h3>

                <div>
                    <div className="add-container">
                        <label htmlFor={item.id}>Quantity :</label>
                        <button onClick={left} className="left-arrow-input">
                            <svg
                                xmlns="http://www.w3.org/2000/svg"
                                viewBox="0 0 24 24"
                            >
                                <title>menu-left</title>
                                <path d="M14,7L9,12L14,17V7Z" />
                            </svg>
                        </button>
                        <input
                            value={input}
                            onChange={(e) => setInput(e.target.value * 1)}
                            type="text"
                            inputMode="numeric"
                            name="quantity"
                            id={item.id}
                        ></input>
                        <button onClick={right} className="right-arrow-input">
                            <svg
                                xmlns="http://www.w3.org/2000/svg"
                                viewBox="0 0 24 24"
                            >
                                <title>menu-right</title>
                                <path d="M10,17L15,12L10,7V17Z" />
                            </svg>
                        </button>
                        <button
                            className="add-button"
                            onClick={() => addItemInList(item, input)}
                        >
                            {action}
                        </button>
                    </div>
                    <p className="price">
                        {input} unit x {item.price}€ ={' '}
                        {Math.round(item.price * input * 100) / 100} €
                    </p>
                </div>
            </div>
        </>
    )
}
