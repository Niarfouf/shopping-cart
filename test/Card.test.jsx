import { vi } from 'vitest'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import Card from '../src/Components/Card'
const item1 = {
    title: 'testTitle1',
    image: 'https://fakestoreapi.com/img/61U7T1koQqL._AC_SX679_.jpg',
    id: 0,
    price: 5.5,
}
const item2 = {
    title: 'testTitle2',
    image: 'https://fakestoreapi.com/img/71kWymZ+c+L._AC_SX679_.jpg',
    id: 1,
    price: 10,
}

describe('Card', () => {
    it('render match snapshot 1', () => {
        const { container } = render(
            <Card
                item={item1}
                addItemInList={() => {}}
                quantity={2}
                action={'test1'}
            />
        )

        expect(container).toMatchSnapshot()
    })
    it('render match snapshot 2', () => {
        const { container } = render(
            <Card
                item={item2}
                addItemInList={() => {}}
                quantity={4}
                action={'test2'}
            />
        )

        expect(container).toMatchSnapshot()
    })

    it('left button should change input when clicked', async () => {
        const user = userEvent.setup()
        render(
            <Card
                item={item1}
                addItemInList={() => {}}
                quantity={2}
                action={'Add'}
            />
        )

        const button = screen.getAllByRole('button')

        await user.click(button[0])
        const input = screen.getByDisplayValue('1')
        expect(input).toBeInTheDocument()
    })
    it('right button should change input when clicked', async () => {
        const user = userEvent.setup()
        render(
            <Card
                item={item1}
                addItemInList={() => {}}
                quantity={2}
                action={'Add'}
            />
        )

        const button = screen.getAllByRole('button')

        await user.click(button[1])
        const input = screen.getByDisplayValue('3')
        expect(input).toBeInTheDocument()
    })
    it('typing in input with keyboard should change input', async () => {
        const user = userEvent.setup()
        render(
            <Card
                item={item1}
                addItemInList={() => {}}
                quantity={2}
                action={'Add'}
            />
        )

        const input = screen.getByDisplayValue('2')
        await user.click(input)
        await user.clear(input)
        await user.keyboard('3')
        const newInput = screen.getByDisplayValue('3')
        expect(newInput).toBeInTheDocument()
    })
    it('Add button should call the addItemInList function when clicked', async () => {
        const addItemInList = vi.fn()
        const user = userEvent.setup()
        render(
            <Card
                item={item1}
                addItemInList={addItemInList}
                quantity={2}
                action={'Add'}
            />
        )

        const button = screen.getByRole('button', { name: 'Add' })

        await user.click(button)

        expect(addItemInList).toHaveBeenCalled()
    })
    it('should not call the addItemInList function when not clicked', async () => {
        const addItemInList = vi.fn()

        render(
            <Card
                item={item1}
                addItemInList={addItemInList}
                quantity={2}
                action={'Add'}
            />
        )

        expect(addItemInList).not.toHaveBeenCalled()
    })
})
