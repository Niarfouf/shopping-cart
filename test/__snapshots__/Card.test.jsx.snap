// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card > render match snapshot 1 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="testTitle1"
      class="item-image"
      src="https://fakestoreapi.com/img/61U7T1koQqL._AC_SX679_.jpg"
    />
    <h3
      class="title"
    >
      testTitle1
    </h3>
    <div>
      <div
        class="add-container"
      >
        <label
          for="0"
        >
          Quantity :
        </label>
        <button
          class="left-arrow-input"
        >
          <svg
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              menu-left
            </title>
            <path
              d="M14,7L9,12L14,17V7Z"
            />
          </svg>
        </button>
        <input
          id="0"
          inputmode="numeric"
          name="quantity"
          type="text"
          value="2"
        />
        <button
          class="right-arrow-input"
        >
          <svg
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              menu-right
            </title>
            <path
              d="M10,17L15,12L10,7V17Z"
            />
          </svg>
        </button>
        <button
          class="add-button"
        >
          test1
        </button>
      </div>
      <p
        class="price"
      >
        2
         unit x 
        5.5
        € =
         
        11
         €
      </p>
    </div>
  </div>
</div>
`;

exports[`Card > render match snapshot 2 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="testTitle2"
      class="item-image"
      src="https://fakestoreapi.com/img/71kWymZ+c+L._AC_SX679_.jpg"
    />
    <h3
      class="title"
    >
      testTitle2
    </h3>
    <div>
      <div
        class="add-container"
      >
        <label
          for="1"
        >
          Quantity :
        </label>
        <button
          class="left-arrow-input"
        >
          <svg
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              menu-left
            </title>
            <path
              d="M14,7L9,12L14,17V7Z"
            />
          </svg>
        </button>
        <input
          id="1"
          inputmode="numeric"
          name="quantity"
          type="text"
          value="4"
        />
        <button
          class="right-arrow-input"
        >
          <svg
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              menu-right
            </title>
            <path
              d="M10,17L15,12L10,7V17Z"
            />
          </svg>
        </button>
        <button
          class="add-button"
        >
          test2
        </button>
      </div>
      <p
        class="price"
      >
        4
         unit x 
        10
        € =
         
        40
         €
      </p>
    </div>
  </div>
</div>
`;
